File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/BCECriterionPerform.lua
--
-- Created by yansh on 16 - 6 - 15.
--

require 'nn'

forwardIterations = 10
backwardIterations = 10
forwardTimeoutMillis = 13
backwardTimeoutMillis = 10
inputNum = 100
featureDim = 512
seed = 100

torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

criterion = nn.BCECriterion()
input = torch.Tensor(inputNum, featureDim)
target = torch.Tensor(inputNum, featureDim)

timer = torch.Timer()
for j = 1, forwardIterations do
    criterion:forward(input, target)
end
timeSpend = timer:time().real / forwardIterations
file:write("BCECriterion_forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

timer = torch.Timer()
for j = 1, backwardIterations do
    criterion:backward(input, target)
end
timeSpend = timer:time().real / backwardIterations
file:write("BCECriterion_backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

file:close()

File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/BCECriterionPerform.scala
package com.intel.analytics.dllib.lib.nn.performTest

import com.intel.analytics.dllib.lib.nn.BCECriterion
import com.intel.analytics.dllib.lib.tensor.{torch}
import org.scalatest.{BeforeAndAfter, FlatSpec}
import com.intel.analytics.dllib.lib.tensor.RandomGenerator._

/**
  * Created by yao on 5/19/16.
  */
class BCECriterionPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val ForwardTimeoutMillis = 20
  val BackwardTimeoutMillis = 20
  val InputNum = 100
  val FeatureDim = 512
  val Seed = 100

  RNG.setSeed(Seed)
  val input = torch.Tensor[Float](InputNum, FeatureDim).rand()
  val target = torch.Tensor[Float](InputNum, FeatureDim).rand()
  val criterion = new BCECriterion[Float]()

  "BCECriterion " should "forward with a good performance" in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testCriterionForwardPerform(input, target, criterion, ForwardIterations)
    //assert(timeMillis < ForwardTimeoutMillis)
  }

  it should " backward with a good performance" in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testCriterionBackwardPerform(input, target, criterion, BackwardIterations)
    //assert(timeMillis < BackwardTimeoutMillis)
  }
}
